src/docfx/lib/HtmlUtility.cs

```diff
@@ -163,6 +164,56 @@ public static string TransformLinks(string html, Func<string, string> transform)
             return result.ToString();
         }
 
+        public static (List<(string uid, int line, int column)>, string) TransformXrefs(string html, int startLine, Func<string, (Error error, string href, string display, Document file)> transform)
+        {
+            var errors = new List<(string uid, int line, int column)>();
+
+            // Fast pass it does not have <xref> tag
+            if (!(html.Contains("<xref", StringComparison.OrdinalIgnoreCase) && html.Contains("href", StringComparison.OrdinalIgnoreCase)))
+            {
+                return (errors, html);
+            }
+
+            var doc = new HtmlDocument();
+            doc.LoadHtml(html);
+
+            var result = new StringBuilder(html.Length + 64);
+            var currentLine = startLine;
+            foreach (var node in doc.DocumentNode.Descendants())
+            {
+                if (node.Name != "xref")
+                {
+                    continue;
+                }
+
+                var xref = node.Attributes["href"];
+                if (xref is null)
+                {
+                    continue;
+                }
+
+                var logWarning = node.GetAttributeValue("data-throw-if-not-resolved", null);
+                var rawSource = node.GetAttributeValue("data-raw-source", null);
+                var rawHtml = node.GetAttributeValue("data-raw-html", null);
+                var raw = HttpUtility.HtmlDecode(!string.IsNullOrEmpty(rawHtml) ? rawHtml : rawSource);
+                var (_, resolvedHref, display, _) = transform(xref.Value);
+                if (string.IsNullOrEmpty(resolvedHref))
+                {
+                    if (string.Compare(logWarning, "False", StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        errors.Add((xref.Value, currentLine, s_getValueStartIndex(xref)));
+                    }
+                    result.Append($"{raw}");
+                }
+                else
+                {
+                    result.Append($"<a href='{resolvedHref}'>{display}</a>");
+                }
+                currentLine += 1;
```

Comment:
Is there a more accurate way to calculate line? This may be wrong if `continue` is hit, or multiple `<xref>` in the same like.

Reply:
For the `continue` case, you are right, updated the test accordingly.
Processing node by node here, I assume each node should be 1 line.
For multiple `<xref>` in the same line, they should have the same line number

Reply:
good point, updated test